src/main/scala/com/fulcrumgenomics/commons/async/AsyncRunnable.scala

```diff
@@ -0,0 +1,168 @@
+  private val threadsCreated = new AtomicInteger(0)
+
+  /** Generates a name for thread that will run the given runnable. */
+  private[AsyncRunnable] def nextName(asyncRunnable: AsyncRunnable): String = {
+    asyncRunnable.getClass.getSimpleName + this.threadsCreated.incrementAndGet
+  }
+}
+
+/** A trait that can be mixed in to help manage asynchronous execution while being [[Runnable]].
+  *
+  * The `execute()` method should contain the work to be performed asynchronously.
+  *
+  * The `thread()` method can be used to return a [[Thread]] that wraps this [[Runnable]].
+  *
+  * The `start()` method can be used to create and start daemon [[Thread]] that wraps this [[Runnable]].
+  *
+  * The `checkAndRaise()` method should be used by implementing classes to check if the
+  * `execute()`  method has encountered an exception, and if so, it will rethrow the exception as an
+  * [[Error]] or [[RuntimeException]].
+  *
+  * The `tryAndModifyInterruptedException()` method should wrap blocking code in implementing classes
+  * that are not part of the `execute()` method.  This method will throw a [[RuntimeException]] if interrupted while
+  * waiting.
+  */
+trait AsyncRunnable extends Runnable {
+
+  /** A reference to an exception thrown during an asynchronous operation. */
+  private[AsyncRunnable] val _throwable: AtomicReference[Throwable] = new AtomicReference(null)
+
+  /** This signals that the run method has started. */
+  private val startedLatch: CountDownLatch = new CountDownLatch(1)
+
+  /** This signals that the run method has completed, even if an exception occurred. This is needed so the close method
+    * can be sure that the run method is no longer executing. */
+  private val doneLatch: CountDownLatch = new CountDownLatch(1)
+
+  final def run(): Unit = {
+    val outer = this
+    startedLatch.countDown()
+    try {
+      outer.execute()
+    } catch {
+      case thr: Throwable =>
+        outer._throwable.compareAndSet(null, thr)
+        outer.uponException()
+    } finally {
+      outer.uponFinally()
+      this.doneLatch.countDown()
+    }
+  }
+
+  /** Returns true if the [[run()]] method has started, false otherwise. */
+  def started: Boolean = this.startedLatch.getCount == 0
+
+  /** Returns true if the [[run()]] method has completed, false otherwise. */
+  def done: Boolean = this.doneLatch.getCount == 0
+
+  /** Returns a throwable if an exception occurred in the [[run()]] method, None otherwise. */
+  def throwable: Option[Throwable] = Option(this._throwable.get())
+
+  /** Waits for the [[run()]] method to start. */
+  def awaitStart(): Unit = this.startedLatch.await()
+
+  /** Returns an [[Awaitable]] that completes when the [[run()]] method has started.  Returns the throwable if an
+    * exception was encountered, [[None]] otherwise. */
+  def startAwaitable()(implicit ec: ExecutionContext) : Awaitable[Unit] = Future { this.awaitStart() }
```

Comment:
I'm having trouble coming up with a better name, but to me this sounds imperative like "start the awaitable!".

Reply:
How about `uponStart` and `uponDone`?